src/foam/nanos/medusa/CompactionDAO.js

```diff
@@ -0,0 +1,413 @@
+      `
+    },
+    {
+      name: 'maxWait',
+      class: 'Long',
+      value: 60000,
+      units: 'ms'
+    },
+    {
+      documentation: 'true when blocking for dagger and node setup',
+      name: 'blocking',
+      class: 'Boolean',
+      visibility: 'HIDDEN'
+    }
+  ],
+
+  methods: [
+    {
+      name: 'cmd_',
+      javaCode: `
+      if ( COMPACTION_CMD.equals(obj) ) {
+        ReplayingInfo replaying = (ReplayingInfo) x.get("replayingInfo");
+        if ( replaying.getReplaying() ) {
+          Loggers.logger(x, this).warning("Compaction not allowed during replay");
+          throw new RuntimeException("Compaction not allowed during Replay");
+        }
+
+        ClusterConfigSupport support = (ClusterConfigSupport) x.get("clusterConfigSupport");
+        ClusterConfig config = support.getConfig(x, support.getConfigId());
+        if ( ! config.getIsPrimary() ) {
+          throw new RuntimeException("Compaction not allowed from Secondaries");
+        }
+
+        // TODO: start in own thread
+        execute(x);
+        return obj;
+      }
+      return getDelegate().cmd_(x, obj);
+      `
+    },
+    {
+      name: 'maybeBlock',
+      args: 'X x',
+      javaCode: `
+      return getBlocking();
+      `
+    },
+    {
+      name: 'execute',
+      args: 'X x',
+      javaCode: `
+      Logger logger = Loggers.logger(x, this);
+      logger.info("start");
+      ClusterConfigSupport support = (ClusterConfigSupport) x.get("clusterConfigSupport");
+
+      try {
+        // block
+        logger.info("block");
+        setBlocking(true);
+
+        // wait
+        logger.info("in-flight");
+        Object response = getDelegate().cmd(MedusaBroadcast2NodesDAO.IN_FLIGHT_CMD);
+        if ( response != null ) {
+          Long inFlight = (Long) response;
+          if ( inFlight > 0L ) {
+            // TODO: wait
+            logger.warning("in-flight", inFlight);
+          }
+        }
+
+        // nodes
+        logger.info("nodes");
+        rollNodes(x);
+
+        // dagger
+        logger.info("dagger");
+        long oldIndex = dagger(x);
+
+        // unblock
+        logger.info("unblock");
+        setBlocking(false);
+        super.unblock(x);
+
+        // compaction
+        logger.info("compaction");
+        compaction(x);
+
+        // purge
+        logger.info("purge");
+        purge(x, oldIndex);
+
+        logger.info("end");
+      } catch (Throwable t) {
+        logger.error(t);
+      }
+      `
+    },
+    {
+      name: 'rollNodes',
+      args: 'X x',
+      javaCode: `
+      final Logger logger = Loggers.logger(x, this);
+      // logger.info("rollNodes");
+
+      final ClusterConfigSupport support = (ClusterConfigSupport) x.get("clusterConfigSupport");
+      final ClusterConfig myConfig = support.getConfig(x, support.getConfigId());
+
+      AssemblyLine line = new AsyncAssemblyLine(x, null, support.getThreadPoolName());
+      final Set replies = new HashSet();
```

Comment:
Should it use thread-safe set? Collections.synchronizedSet(new HashSet<>());

Reply:
Refactored to use map.